src/foam/core/CompoundContextAgency.js

```diff
@@ -0,0 +1,91 @@
+/**
+ * @license
+ * Copyright 2019 The FOAM Authors. All Rights Reserved.
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+foam.CLASS({
+  package: 'foam.core',
+  name: 'CompoundContextAgency',
+  javaImplements: [
+    'foam.core.Agency',
+    'foam.core.ContextAgent'
+  ],
+
+  javaImports: [
+    'java.util.ArrayList'
+  ],
+
+  properties: [
+    {
+      name: 'agents',
+      class: 'List',
+      javaType: 'ArrayList<Runnable>',
+      javaFactory: `return new ArrayList();`
+    }
+  ],
+
+  methods: [
+    {
+      name: 'execute',
+      args: [
+        { name: 'x', type: 'X' }
+      ],
+      javaCode: `CompoundException e = new CompoundException();
+for ( Runnable agent : getAgents() ) {
+  try {
+    agent.run();
+  } catch (Throwable t) {
+    e.exceptions.add(t);
+  }
+}
+if ( e.exceptions.size() != 0 ) {
```

Comment:
Again, better to let the CompoundException decide this. Here just call e.maybeThrow(); or e.rethrow(); and let it count it's exceptions and decide if it wants to throw or not.